src/modules/iotjs_module_mqtt.c

```diff
@@ -382,37 +386,48 @@ JS_FUNCTION(MqttDeserialize) {
   int qos = 0;
 
   MQTTString topic;
-  char* payload;
-  int payload_len;
+  unsigned char* payload;
+  int payload_size;
+  unsigned char* buf = (unsigned char*)iotjs_bufferwrap_buffer(wrap);
+  int buf_size = (int)iotjs_bufferwrap_length(wrap);
 
-  int r = MQTTDeserialize_publish(&dup, (int*)&qos, &retain, &msgId, &topic,
-                                  (unsigned char**)&payload, (int*)&payload_len,
-                                  (unsigned char*)iotjs_bufferwrap_buffer(wrap),
-                                  (int)iotjs_bufferwrap_length(wrap));
+  int r = MQTTDeserialize_publish(&dup, &qos, &retain, &msgId, &topic, &payload,
+                                  &payload_size, buf, buf_size);
   if (r != 1) {
     return JS_CREATE_ERROR(COMMON, "failed to deserialize publish message.");
   }
+  int header_size = (int)(payload - buf);
+  int payload_real_size = buf_size - header_size;
+  int payload_missing_size = payload_size - payload_real_size;
 
   jerry_value_t msg = jerry_create_object();
-  iotjs_jval_set_property_jval(msg, "id", jerry_create_number(msgId));
-  iotjs_jval_set_property_jval(msg, "qos", jerry_create_number(qos));
-  iotjs_jval_set_property_jval(msg, "dup", jerry_create_boolean(dup));
-  iotjs_jval_set_property_jval(msg, "retain", jerry_create_boolean(retain));
+  iotjs_jval_set_property_number(msg, "id", msgId);
+  iotjs_jval_set_property_number(msg, "qos", qos);
+  iotjs_jval_set_property_number(msg, "headerSize", header_size);
+  iotjs_jval_set_property_number(msg, "payloadRealSize", payload_real_size);
+  iotjs_jval_set_property_number(msg, "payloadMissingSize",
+                                 payload_missing_size);
+  iotjs_jval_set_property_number(msg, "payloadSize", payload_size);
+  iotjs_jval_set_property_boolean(msg, "dup", dup);
+  iotjs_jval_set_property_boolean(msg, "retain", retain);
```

Comment:
That would be better if these keys are in magic string.

Reply:
What I did mean is here: https://github.com/Rokid/ShadowNode/blob/master/src/iotjs_magic_strings.h :)